project1/server/server_tasks/agent_task.cpp

```diff
@@ -1,25 +1,30 @@
 #include "agent_task.h"
 
+#include <utility>
+
 namespace server_tasks {
 
-    void AgentTask::SetClientFD(int fd) {
-        client_fd_ = fd;
-    }
+    AgentTask::AgentTask(int id, std::shared_ptr<CommandIDs> cmds,
+                         std::shared_ptr<server::file_handler::FileAccessManager> read_mgr,
+                         std::shared_ptr<server::file_handler::FileAccessManager> write_mgr)
+            : client_fd_(id), active_commands_(std::move(cmds)),
+              read_manager_(std::move(read_mgr)), write_manager_(std::move(write_mgr)) {}
 
-    void AgentTask::SetActiveCommands(std::shared_ptr<CommandIDs>cmds) {
-        active_commands_ = cmds;
-    }
-
-    void AgentTask::SetFileAccessManagers(std::shared_ptr<server::file_handler::FileAccessManager> read_mgr,
-                                          std::shared_ptr<server::file_handler::FileAccessManager> write_mgr) {
-        read_manager_ = read_mgr;
-        write_manager_ = write_mgr;
-    }
+    AgentTask::AgentTask(int id, std::shared_ptr<CommandIDs> cmds)
+    : client_fd_(id), active_commands_(cmds) {}
 
     thread_pool::Task::Status AgentTask::SetUp() {
-        // give the agent a unique file handler with the shared access managers
-        auto fh = std::make_unique<server::file_handler::ThreadSafeFileHandler>(read_manager_,write_manager_);
-        agent_ = new server::Agent(client_fd_, std::move(fh),active_commands_);
+
+        // if this is an agent for a normal command, these members should be initialized
+        if (read_manager_ && write_manager_) {
+            // give the agent a unique file handler with the shared access managers
+            auto fh = std::make_unique<server::file_handler::ThreadSafeFileHandler>(read_manager_, write_manager_);
+            agent_ = new server::Agent(client_fd_, std::move(fh), active_commands_);
```

Comment:
Don't use `new` and `delete`. Use `std::unique_ptr` instead.

Reply:
Apologies for not fixing this last time, I had some trouble figuring out syntax.

Reply:
Got it.